### PR TITLE
VexFlow in Node JS (with examples). Improve Renderer constructor and types.

### PR DESCRIPTION
**Pull request by ronyeh**
Inspired by the PR from @aleen42  https://github.com/0xfe/vexflow/pull/1045

I will continue to test this PR, but I would suggest this approach to patching `renderer.ts`.

Since this PR was inspired by @aleen42, I would like to give credit where it's due. I'm happy if @aleen42 takes the `renderer.ts` from this PR and submits it. 

This PR also adds a `demos/` folder, where we can hold representative examples of "how to use VexFlow."

### CODE REVIEW

**Comment by ronyeh** [comment]
`npm run test:reference` shows no visual diffs from master revision 927f1c140707213fee92636244e9f3ab3333fee7

The two files in the `demos/` folder show that this PR allows the VexFlow library to be accessed in Node via `require()`.

Ready for code review.

**Comment by aleen42** [comment]
Ater merging this PR, I will rebase my branch.

**Comment by ronyeh** [comment]
Thanks for pointing out the issue and suggesting the change!

**Comment by aleen42** [comment]
@ronyeh In my opinion, it is suggested that creating PR to my branch at first, and then give a combined PR for the implementation and the demo together towards the original repo.

**Comment by ronyeh** [comment]
Hello! I think I've spent enough time on this already.... But I would totally support it if you submit an identical PR. Just grab the 5 changed files from here: https://github.com/ronyeh/vexflow/tree/examples

This is the main file: https://raw.githubusercontent.com/0xfe/vexflow/9bc068976e6b523cfcfee2f0570c37cbd656017d/src/renderer.ts

If you submit a new PR with no diffs from mine, I would be happy to close this PR off vote for yours to be merged into the main repo. That way you can be officially included in the list of vexflow contributors.

**Comment by aleen42** [comment]
I'll pick partials of your commits, and skip the changes of `rerender.ts`, because there is some conflicts.

**Comment by ronyeh** [comment]
I'm just saying that I have fully tested and would 100% endorse exactly this patch (without diffs from this PR). If your PR takes a different approach, you'll have to do the work of testing and convincing this repo's owner to accept your patch instead. For example, my approach avoided the || 0 which seemed to confuse some of us since it is not immediately apparent what it does.

**Comment by aleen42** [comment]
Of course, I have tested your provided demos in my approach since two examples were provided first in: https://github.com/0xfe/vexflow/pull/1045#issuecomment-873714575 and https://github.com/0xfe/vexflow/pull/1045#issuecomment-873905207. If there is any confusion during merging #1045, you can review it in that PR.

File: Gruntfile.js [hunk 1]
```diff
@@ -37,7 +37,7 @@ module.exports = (grunt) => {
         library: libraryName,
         libraryTarget: 'umd',
         libraryExport: 'default',
-        globalObject: `typeof window !== 'undefined' ? window : this`,
+        globalObject: `typeof window !== 'undefined' ? window : typeof globalThis !== 'undefined' ? globalThis : this`,
```
Comment (ronyeh) [inline Gruntfile.js:40]: @0xfe I have tested this with legacy Node JS which does not include globalThis. Guarding it with typeof prevents any errors, and enables VexFlow to be imported as a JS module in Node.

File: src/canvascontext.ts [hunk 2]
```diff
@@ -14,22 +14,24 @@ export class CanvasContext implements RenderContext {
...
     return 32767; // Chrome/Firefox. Could be determined more precisely by npm module canvas-size.
   }
 
   static SanitizeCanvasDims(width: number, height: number): [number, number] {
-    if (Math.max(width, height) > this.CANVAS_BROWSER_SIZE_LIMIT) {
-      warn('Canvas dimensions exceed browser limit. Cropping to ' + this.CANVAS_BROWSER_SIZE_LIMIT);
-      if (width > this.CANVAS_BROWSER_SIZE_LIMIT) {
-        width = this.CANVAS_BROWSER_SIZE_LIMIT;
-        // note: Math.min return 0 for undefined, NaN for null. Would change inputs.
+    const limit = this.CANVAS_BROWSER_SIZE_LIMIT;
```
Comment (ronyeh) [inline src/canvascontext.ts:27]: This is just a small fix to improve readability (& reduce duplication).

File: src/renderer.ts [hunk 3]
```diff
@@ -13,8 +13,8 @@ export type ContextBuilder = typeof Renderer.getSVGContext | typeof Renderer.get
  * Support Canvas & SVG rendering contexts.
  */
 export class Renderer {
-  protected elementId?: string | HTMLCanvasElement;
-  protected element: HTMLCanvasElement;
+  protected elementId?: string;
+  protected element: HTMLCanvasElement | HTMLDivElement;
```
Comment (ronyeh) [inline src/renderer.ts:17]: element can be either a canvas or a div, because SVG contexts take a div.

**Review by ronyeh — COMMENTED** [review]
@0xfe I believe this is ready for review. Tested and no visual diffs.

This PR also improves the types in Renderer, because the element property can be either a canvas or a div.

File: Gruntfile.js (hunk 1 above)
Comment (0xfe) [reply Gruntfile.js:40]: Can you add a comment here (with a bit of detail) around what's going on here?

File: src/canvascontext.ts (hunk 2 above)
Comment (0xfe) [reply src/canvascontext.ts:27]: Thanks for this 👍

File: src/renderer.ts [hunk 4]
```diff
@@ -145,23 +145,34 @@ export class Renderer {
...
-  constructor(elementId: string | HTMLCanvasElement, backend: number) {
-    if (!elementId) {
+  /**
+   * @param id_canvas_div the first parameter can be:
+   *   - a string element ID (of a canvas or div element)
+   *   - a canvas element
+   *   - a div element, which will contain the SVG output
+   * @param backend Renderer.Backends.CANVAS or Renderer.Backends.SVG
+   */
+  constructor(id_canvas_div: string | HTMLCanvasElement | HTMLDivElement, backend: number) {
```
Comment (0xfe) [inline src/renderer.ts:155]: Call this `canvasId` please.

**Comment by ronyeh** [comment]
Addressed code review comments. Ran the visual tests against most recent master with 0 diffs.

**Comment by ronyeh** [comment]
Tested on September 4th. Visual regression tests show zero diffs relative to master at b37ddd7a228c69b165c96f4431598b703585fc37. There are new images due to the new `renderer_tests.ts`.

This PR improves the Renderer constructor by **explicitly declaring** that the first argument can be a `string | HTMLCanvasElement | HTMLDivElement`.

I added RendererTests, since passing in a `<canvas>` or `<div>`  had not been covered in our unit tests (even though the [VexFlow README](https://github.com/0xfe/vexflow#using-the-native-api) shows that a `<div>` can be passed directly to Renderer's constructor).


This PR also improves the Gruntfile by adding `globalThis` as one of the options for the globalObject. This allows VexFlow to be used by JS modules in Node JS (where `this` is undefined, but `globalThis` points to the Node global object).

@0xfe I think this one is ready! The only new changes from your last review are the new `renderer_tests.ts` file, and Gruntfile comments and a renamed variable to address your comments.

**Comment by 0xfe** [comment]
Thanks Ron. Can you fix that one conflict from the last merge please?

**Comment by ronyeh** [comment]
Merge conflict resolved.

**Comment by 0xfe** [comment]
Thanks. Looks good.

**Comment by aleen42** [comment]
👍🏼

**Pull request merged**